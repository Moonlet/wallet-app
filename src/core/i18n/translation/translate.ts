import { ITranslations, ITranslation } from './types';

/**
 * Plural specs:
 * http://www.unicode.org/cldr/charts/28/supplemental/language_plural_rules.html
 */
export interface ITranslationParams {
    [key: string]: any;
}

export type Language = 'en';

const store: {
    language: Language;
    translations: ITranslations;
} = {
    language: undefined as any,
    translations: undefined as any
};

export const loadTranslations = (language: Language) => {
    let promise;
    switch (language) {
        case 'en':
        default:
            promise = import('./locales/en' /* webpackChunkName = "translations/en" */);
    }

    return promise.then(mod => {
        store.language = language;
        store.translations = mod.translation;
        return mod.translation;
    });
};

const getKeyConfig = (key: string): ITranslation | string => {
    const keyParts = key.split('.');
    let config = store?.translations?.texts as any;

    for (const keyPart of keyParts) {
        if (config && config[keyPart]) {
            config = config[keyPart];
        } else {
            config = undefined;
            break;
        }
    }

    return config;
};

const format = (text: string, params?: ITranslationParams): string => {
    if (params) {
        for (const key in params) {
            if (params.hasOwnProperty(key)) {
                text = text.replace(new RegExp('{{' + key + '}}', 'g'), String(params[key]));
            }
        }
    }

    return text;
};

export const translate = (key: string, params?: ITranslationParams, count: number = 0) => {
    const translationConfig = getKeyConfig(key);

    if (translationConfig) {
        if (typeof translationConfig === 'string') {
            return format(translationConfig, params);
        }

        const plural = store.translations.plural(count, translationConfig.ordinal);
        const forms: any = translationConfig.forms;
        if (translationConfig.forms && forms[plural]) {
            return format(forms[plural], params);
        } else {
            return format(translationConfig.text, params);
        }
    } else {
        return '';
    }
};
