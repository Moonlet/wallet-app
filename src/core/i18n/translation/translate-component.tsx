import React from 'react';
import { Text, ITextProps } from '../../../library';
import { ITranslationParams, translate } from './translate';
import { withTheme } from '../../theme/with-theme';

interface IProps {
    text: string;
    params?: ITranslationParams;
    count?: number;
}

export class TranslateComponent extends React.Component<IProps & ITextProps> {
    public render() {
        return (
            <Text {...this.props}>
                {translate(this.props.text, this.props.params, this.props.count)}
            </Text>
        );
    }
}

export const Translate = withTheme(() => ({}))(TranslateComponent);
