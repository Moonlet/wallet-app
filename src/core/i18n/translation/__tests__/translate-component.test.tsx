import React from 'react';
import { TranslateComponent, Translate } from '../translate-component';
import { shallow } from 'enzyme';
import { translate } from '../translate';

jest.mock('../translate', () => ({
    translate: jest.fn(() => 'TEST RESULT')
}));

export default describe('Translate component', () => {
    test('it renders corectly', () => {
        const T = TranslateComponent as any;
        expect(shallow(<T text="TEST" />).debug()).toMatchSnapshot();
        expect(translate).toHaveBeenCalledWith('TEST', undefined, undefined);

        (translate as any).mockClear();
        shallow(<T text="TEST" params={'PARAMS'} count={123} />);
        expect(translate).toHaveBeenCalledWith('TEST', 'PARAMS', 123);

        expect(
            shallow(<Translate text="TEST" params={'PARAMS'} count={123} />).debug()
        ).toMatchSnapshot();
    });
});
