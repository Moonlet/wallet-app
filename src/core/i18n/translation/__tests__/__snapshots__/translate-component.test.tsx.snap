// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translate component it renders corectly 1`] = `
"<withTheme(TextComponent) text=\\"TEST\\">
  TEST RESULT
</withTheme(TextComponent)>"
`;

exports[`Translate component it renders corectly 2`] = `"<TranslateComponent text=\\"TEST\\" params=\\"PARAMS\\" count={123} styles={{...}} theme={{...}} />"`;
